tokio/src/sync/watch.rs

```diff
@@ -266,79 +228,96 @@ impl<T: Clone> Receiver<T> {
     /// async fn main() {
     ///     let (tx, mut rx) = watch::channel("hello");
     ///
-    ///     let v = rx.recv().await;
-    ///     assert_eq!(v, "hello");
-    ///
     ///     tokio::spawn(async move {
     ///         tx.send("goodbye").unwrap();
     ///     });
     ///
-    ///     // Waits for the new task to spawn and send the value.
-    ///     let v = rx.recv().await;
-    ///     assert_eq!(v, "goodbye");
+    ///     assert!(rx.changed().await.is_ok());
+    ///     assert_eq!(*rx.borrow(), "goodbye");
     ///
-    ///     let v = rx.recv().await;
-    ///     assert_eq!(v, "goodbye");
+    ///     // The `tx` handle has been dropped
+    ///     assert!(rx.changed().await.is_err());
     /// }
     /// ```
-    pub async fn recv(&mut self) -> T {
-        poll_fn(|cx| {
-            let v_ref = ready!(self.poll_recv_ref(cx));
-            Poll::Ready((*v_ref).clone())
-        })
-        .await
+    pub async fn changed(&mut self) -> Result<(), error::RecvError> {
+        // In order to avoid a race condition, we first request a notification
+        // **then** check the current value's version. If a new version exists,
+        // the notification request is dropped.
+        let notified = self.shared.notify_rx.notified();
```

Comment:
The key is that calling `notified()` inserts a waiter into the notify internal state. If another thread is racing to update the watch state and notify, this could result in an extra task wake instead of a lost task wake.

Consider:

**Thread A**
* Checks the watch state, notices no change
* Awaits on `notified()`.

**Thread B**
* Updates watch state
* Notifies

There is a potential for all of thread B's steps to happen between the two thread A steps. This would result in a missed update. Instead, by registering interest in notification **before** checking the watch state, the race is avoided.

Reply:
> [...] inserts a waiter into the notify internal state [...]

I think @hawkw's concern is that the current `notified` implementation doesn't do that:

https://github.com/tokio-rs/tokio/blob/b8b069320bd408c511bd5ae4aad78a7103af964c/tokio/src/sync/notify.rs#L206-L217

Reply:
:upside_down_face: Ah yes... good catch.

I should make sure there is a loom test catching this case.

Reply:
Okay, I'm glad I wasn't just losing my mind. :)

Reply:
Ok, I fixed this & added a test.